lib/charms/oathkeeper/v0/auth_proxy.py

```diff
@@ -0,0 +1,448 @@
+
+```yaml
+requires:
+  auth-proxy:
+    interface: auth_proxy
+    limit: 1
+```
+
+Then, to initialise the library:
+```python
+from charms.oathkeeper.v0.auth_proxy import AuthProxyConfig, AuthProxyRequirer
+
+AUTH_PROXY_ALLOWED_ENDPOINTS = ["welcome", "about/app"]
+AUTH_PROXY_HEADERS = ["X-User", "X-Some-Header"]
+
+class SomeCharm(CharmBase):
+    def __init__(self, *args):
+        # ...
+        self.auth_proxy = AuthProxyRequirer(self, self._auth_proxy_config)
+
+        @property
+        def _auth_proxy_config(self) -> AuthProxyConfig:
+            return AuthProxyConfig(
+                protected_urls=self.external_url,
+                allowed_endpoints=AUTH_PROXY_ALLOWED_ENDPOINTS,
+                headers=AUTH_PROXY_HEADERS
+            )
+
+        def _on_ingress_ready(self, event):
+            self.external_url = "https://example.com"
+            self._configure_auth_proxy()
+
+        def _configure_auth_proxy(self):
+            self.auth_proxy.update_auth_proxy_config(auth_proxy_config=self._auth_proxy_config)
+```
+"""
+
+import json
+import logging
+import re
+from dataclasses import asdict, dataclass, field
+from typing import Dict, List, Mapping, Optional
+
+import jsonschema
+from ops.charm import CharmBase, RelationChangedEvent, RelationCreatedEvent, RelationDepartedEvent
+from ops.framework import EventBase, EventSource, Handle, Object, ObjectEvents
+from ops.model import Relation, TooManyRelatedAppsError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "0e67a205d1c14d7a86d89f099d19c541"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+RELATION_NAME = "auth-proxy"
+INTERFACE_NAME = "auth_proxy"
+
+logger = logging.getLogger(__name__)
+
+ALLOWED_HEADERS = ["X-User"]
+
+url_regex = re.compile(
+    r"(^http://)|(^https://)"  # http:// or https://
+    r"(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|"
+    r"[A-Z0-9-]{2,}\.?)|"  # domain...
+    r"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})"  # ...or ip
+    r"(?::\d+)?"  # optional port
+    r"(?:/?|[/?]\S+)$",
+    re.IGNORECASE,
+)
+
+AUTH_PROXY_REQUIRER_JSON_SCHEMA = {
+    "$schema": "http://json-schema.org/draft-07/schema",
+    "$id": "https://canonical.github.io/charm-relation-interfaces/interfaces/auth-proxy/schemas/requirer.json",
+    "type": "object",
+    "properties": {
+        "protected_urls": {"type": "array", "default": None, "items": {"type": "string"}},
+        "allowed_endpoints": {"type": "array", "default": [], "items": {"type": "string"}},
+        "headers": {
+            "type": "array",
+            "default": ["X-User"],
+            "items": {
+                "enum": ALLOWED_HEADERS,
+                "type": "string",
+            },
+        },
+    },
+    "required": ["protected_urls", "allowed_endpoints", "headers"],
+}
```

Comment:
I think that we should start using pydantic to validate the relation databags. We don't have to do it in this PR, but let's create an issue to keep track of it

Reply:
Ok, I created an issue